plugins/libplugin-pay.c

```diff
@@ -79,6 +83,52 @@ struct payment *payment_root(struct payment *p)
 		return payment_root(p->parent);
 }
 
+static char *
+paymod_log_header(const tal_t *ctx, struct payment *p)
+{
+	struct payment *root = payment_root(p);
+	/* We prefer to show the command ID here since it is also known
+	 * by `lightningd`, so in theory it can be used to correlate
+	 * debugging logs between the main `lightningd` and whatever
+	 * plugin is using the paymod system.
+	 * We only fall back to a unique id per root payment if there
+	 * is no command with an id associated with this payment.
+	 */
+	if (root->cmd && root->cmd->id)
+		return tal_fmt(ctx, "cmd %"PRIu64" partid %"PRIu32,
+			       *root->cmd->id, p->partid);
+	else
+		return tal_fmt(ctx, "id %"PRIu64" partid %"PRIu32,
+			       root->id, p->partid);
+}
```

Comment:
This results in an extra allocation for the string, where it might be preferable to integrate this into the `paymod_log` logic, which is still rather small.

Reply:
We could have a simple tuple returned with the components:

```
struct paymod_log_meta {
    const char *type;
    const u64 payment_id;
    const u32 part_id;
}
```

And then pass that to the `plugin_log` function. If we return this by value rather than by reference it should get allocated on the caller stack (`paymod_log` or `paymod_err`) and we save the extra cost of allocating.